### DISCUSSION TITLE
Support RISC and CAEP events / Shared Signals and Events

### DISCUSSION THREAD

**Discussion by sandhose** [keycloak/keycloak#14217] [Ideas]
To quote the OpenID Shared Signals and Events working group page:

> The Shared Signals and Events (SSE) Framework improves API efficiency and security by providing privacy-protected, secure webhooks. It is in use by some of the largest cloud services to communicate security alerts and status changes of users, continuously and securely to prevent and mitigate security breaches. It is currently leveraged by two applications – the Continuous Access Evaluation Protocol (CAEP) and Risk Incident Sharing and Coordination (RISC) to achieve this result.

This recently gained some traction, with the OpenID RISC Profile Secification 1.0 being adopted as an OpenID implementer's draft: https://openid.net/2022/08/27/second-implementers-draft-of-risc-profile-approved/

This website from Cisco and the OpenID foundation does a great job at explaining what it is: https://sharedsignals.guide

Conceptually, it is a framework send security events between parties. There are currently two specs defining types of events:

 - RISC (Risk Incident Sharing and Coordination), which includes events about the account security, like:
   - when an account is deactivated/re-activated
   - when an account was compromised
   - when an identifier changed or was recycled
   - when an account is being recovered
 - CAEP (Continuous Access Evaluation Protocol), which includes events about the level of access of a session, like
   - when a session is revoked
   - when a claim changed
   - when the ACR level changed
   - when a credential changed

I think it would make sense for Keycloak to support at least some of those events, both as a transmitter (the party sending those events) and as a receiver (the party getting those events). Personally, I'm interested with Keycloak sending those events, but receiving them would help for example to use [Google's Cross-Account Protection](https://developers.google.com/identity/protocols/risc) (which is their marketing name for RISC events)

**Comment by SteveBorger** [comment]
I'd like to see this get pushed as well. Apple now require Shared Signals Framework to support third party IDPs

**Reply by james-callahan** [reply]
I'd love for keycloak to at least stub out an SSF implementation to fool the Apple "Set up your custom identity provider form" to let me click "Continue". It currently doesn't let you proceed without providing an SSF Config URL

Apple docs: https://support.apple.com/en-au/guide/apple-business-manager/axmfcab66783/web

**Reply by willgentsch** [reply]
Signed up to Github just to +1 Steve's reply.

Being able to use KeyCloak with Apple School Manager would be a complete game-changer!

**Reply by embesozzi** [reply]
I believe that  that Keycloak supporting the reception of events from an external IdP would be great. For instance, in a Social login scenario using a Digital Wallet, perhaps Keycloak could add more friction to the login process if it receives RISC events related to that Identity

**Reply by Theschelf** [reply]
+1 for Apple School/Business Manager support

**Comment by thomasdarimont** [comment]
There is also https://caep.dev which provides mock backends to implement custom transmitters.

One can create a receiver for testing here: https://caep.dev/receiver/streams

**Comment by thomasdarimont** [comment]
I started to work on the SSF support for Keycloak here: https://github.com/thomasdarimont/keycloak/tree/poc/shared-signals

The relavant (PoC) code can be found here: https://github.com/thomasdarimont/keycloak/tree/poc/shared-signals/services/src/main/java/org/keycloak/protocol/ssf

I'm currently focussing on the transmitter part of the spec to let Keycloak propagate security relevant events as SETs (security event tokens). This involves, to expose transmitter metadata, offer ways to manage (CRUD) streams and subjects, as well as propagating events via push / pull mechanisms.

After that we can add support to let Keycloak act as SSF receiver.

My current implementation is still very early and requires some additional infrastructure (reverse proxy) to be used with tools like https://caep.dev - the reason for this is that the spec mandates to expose the .well-known endpoint with the transmitter metadata at the root if the domain. Instead of https://id.example.org/realms/demo/.well-known/ssf-configuration, we have to use https://id.example.org/.well-known/ssf-configuration or https://id.example.org/.well-known/ssf-configuration/demo.

Disclaimer: I also work on the OpenID Conformance Tests for the SSF Framework and occasionaly work on the Keycloak side of things to have another "reference" to "test" the tests during development.

**Reply by james-callahan** [reply]
@thomasdarimont have you had any further progress on this since January? Any roadblocks?

**Reply by thomasdarimont** [reply]
Thanks for asking! Adding SSF support to Keycloak directly posed a bit of a challenge, as it was unclear how SSF should be adopted. To get some more experience with this I created a prototype extension for SSF support that allows Keycloak to act as a SSF receiver: https://github.com/identitytailor/keycloak-ssf-support with pluggable security event handlers.
Adding SSF transmitter support comes next.

**Reply by james-callahan** [reply]
@thomasdarimont would you say the prototype was successful? Is it something that you now plan on submitting as a PR to the keycloak repo?
Are you currently looking for feedback on the architecture; the code itself; or something else?

**Reply by james-callahan** [reply]
@thomasdarimont ping? ^

**Comment by james-callahan** [comment]
Looks like authentik is now advertising this as a major win over keycloak.
![image](https://github.com/user-attachments/assets/487c4b0e-2d93-4d4a-9fce-f058d922800f)

https://goauthentik.io/pricing/
![image](https://github.com/user-attachments/assets/0ded70b1-df4a-47fb-9502-2c959d560560)

https://docs.goauthentik.io/docs/add-secure-apps/providers/ssf/

**Comment by Kopf02** [comment]
I would rly appreciate SSF Support. We're using Apple School Manager at our School and Keycloak as SSO and Apple requires SSF for OIDC Login...

**Comment by woaiyoubao** [comment]
Can KeyCloak currently support acting as a SSF transmitter, I want to integrate ABM federated authentication, which requires OIDC and SSF.

**Comment by gitreelike** [comment]
We also need this. Both transmitter and receiver.

@thomasdarimont we'd be happy to contribute in most ways I can currently think of (incl. development, testing, documentation, ...).

**Comment by thomasdarimont** [comment]
FYI I created the following issue https://github.com/keycloak/keycloak/issues/43614 to provide an initial SSF implementation based my previous prototype

**Comment by thomasdarimont** [comment]
I created another issue https://github.com/keycloak/keycloak/issues/43614 to track only the parts with the SSF Transmitter. I also sends a PR that adds SSF transmitter capability to Keycloak and also makes Keycloak compatible with Apple Business Manager: https://github.com/keycloak/keycloak/pull/48256

**Comment by thomasdarimont** [comment]
Note that SSF Transmitter capability is now available in Keycloak as an experimental feature and will be announced soon :) 
This includes compatibility with Apple Business Manager / Apple School Manager.

You can already test it via the nightly Keycloak build (You need to enable the `ssf` feature).

**Reply by bwhdien** [reply]
This will be available as an experimental feature in the next release?

**Reply by thomasdarimont** [reply]
That's the current plan :)

**Comment by idominiki** [comment]
Hi together,
first @thomasdarimont: A big thanks for implementing this feature. 

We're using Keycloak within our UCS@School environment. Yesterday I tried a first step into setting up Keycloak as IdP in Apple School Manager.

The first connection in ASM is working, but after I need to sign-in with an admin IdP user I got an error from Apple 
"Fehlercode -23983: Fehlerbeschreibung: This action could not be completed because of an error. Try signing in again."

Attached my adapter configuration from Keycloak. 

Has anyone get a working Keycloak connection with this new feature?

Thanks!
Dominik
[asm-test.json](https://github.com/user-attachments/files/30738518/asm-test.json)

**Reply by thomasdarimont** [reply]
Please take a look here: https://github.com/keycloak/keycloak/discussions/48840#discussioncomment-17686727
